File: haskell/lib/Tidepool/Ui.hs
{-# LANGUAGE OverloadedStrings #-}

-- | The @Ui@ eDSL — UI described as DATA. R0 is first-order: no monadic
-- sequencing, no @Dialog@ effect, no @uiOf@\/@[form|]@ (those are R1 —
-- see @plans\/harness-r0\/50-ui-edsl\/SPEC.md@). Haskell speaks elicitation
-- semantics (card\/prose\/code\/choice\/text\/badge); Datastar\/HTML
-- vocabulary exists only in the tidepool-web renderer.
--
-- WIRE CONTRACT: the JSON shape here is a cross-language contract PINNED by
-- the Rust mirror @tidepool-harness\/src\/ui.rs@ — its @wire_shape_is_stable@
-- test is canonical. The shape matches exactly: externally tagged by a
-- @"ui"@ field, snake_case tag\/kind values, @options@ as 2-element arrays.
--
-- One thing does NOT match byte-for-byte: object key ORDER. The vendored
-- 'Tidepool.Aeson.Value.object' is backed by @Data.Map.Strict@, which always
-- iterates keys in ascending order — it structurally cannot reproduce the
-- Rust struct's declaration-order (tag-first) field sequence. This is
-- harmless: JSON object member order carries no semantics, and every
-- consumer (serde on the Rust side, this module's own JIT probe) compares
-- structurally, never as a literal string.
module Tidepool.Ui
  ( Ui (..)
  , BadgeKind (..)
    -- * Smart constructors
  , card
  , prose
  , code
  , choice
  , textIn
  , keyedText
  , keyedTextInitial
  , keyedChoice
  , keyedChoiceSelected
  , keyedMultiChoice
  , badge
  ) where

import Prelude
import Data.Text (Text)

import Tidepool.Aeson.Value (ToJSON (..), Value (..), object, (.=))

-- | UI-as-data. Every 'Choice' renders with an open-prose escape path in
-- addition to its options — a renderer-side law (B2), not represented in
-- this type.
data Ui
  = Card   { title :: Text, body :: [Ui] }
  | Prose  { text :: Text }                              -- ^ Markdown.
  | Code   { lang :: Text, source :: Text }               -- ^ Fenced source block.
    -- | @options@ are (option-key, label) pairs. @key@ is the FORM-FIELD key:
    -- 'Just' makes this a field of an enclosing form (renders as a radio group
    -- submitted under @values.<key>@); 'Nothing' is a standalone one-click
    -- choice. (@Tidepool.Form@ sets it; hand-authored @choice@ leaves it off.)
    -- @selected@ is the option-key pre-selected as the radio group's default.
  | Choice { prompt :: Text, options :: [(Text, Text)], key :: Maybe Text, selected :: Maybe Text }
    -- | @initial@ seeds the input's starting text (a draft to edit).
  | TextIn { prompt :: Text, multiline :: Bool, key :: Maybe Text, initial :: Maybe Text }
    -- | Checkboxes over (option-key, label) pairs — a FORM-FIELD widget,
    -- keyed exactly like a keyed 'Choice': submitted under @values.<key>@ as
    -- an ARRAY of the checked option-keys.
  | MultiChoice { prompt :: Text, options :: [(Text, Text)], key :: Maybe Text }
  | Badge  { label :: Text, kind :: BadgeKind }
  deriving (Eq, Show)

-- | Effect-row \/ fan \/ price \/ state chips.
data BadgeKind = EffectRow | Fan | Price | State
  deriving (Eq, Show)

instance ToJSON Ui where
  toJSON u = case u of
    Card t b    -> object [ "ui" .= ("card" :: Text), "title" .= t, "body" .= b ]
    Prose t     -> object [ "ui" .= ("prose" :: Text), "text" .= t ]
    Code l s    -> object [ "ui" .= ("code" :: Text), "lang" .= l, "source" .= s ]
    Choice p os mk msel -> object ([ "ui" .= ("choice" :: Text), "prompt" .= p, "options" .= os ] ++ keyField mk ++ selectedField msel)
    TextIn p m mk mi    -> object ([ "ui" .= ("text_in" :: Text), "prompt" .= p, "multiline" .= m ] ++ keyField mk ++ initialField mi)
    MultiChoice p os mk -> object ([ "ui" .= ("multi_choice" :: Text), "prompt" .= p, "options" .= os ] ++ keyField mk)
    Badge l k   -> object [ "ui" .= ("badge" :: Text), "label" .= l, "kind" .= k ]
    where
      -- Emit the form-field key only when present (matches the Rust mirror's
      -- `skip_serializing_if`, so standalone widgets stay wire-identical).
      keyField = maybe [] (\k -> [ "key" .= (k :: Text) ])
      selectedField = maybe [] (\s -> [ "selected" .= (s :: Text) ])
      initialField = maybe [] (\i -> [ "initial" .= (i :: Text) ])

instance ToJSON BadgeKind where
  toJSON EffectRow = String "effect_row"
  toJSON Fan       = String "fan"
  toJSON Price     = String "price"
  toJSON State     = String "state"

-- ---------------------------------------------------------------------------
-- Smart constructors — ergonomic eval authoring
-- ---------------------------------------------------------------------------

-- | A titled card grouping child elements.
card :: Text -> [Ui] -> Ui
card = Card

-- | Markdown prose.
prose :: Text -> Ui
prose = Prose

-- | A fenced source block: language, then source text.
code :: Text -> Text -> Ui
code = Code

-- | A standalone one-click choice: (option-key, label) options; the renderer
-- adds an open-prose escape unconditionally (B2 — not represented here).
choice :: Text -> [(Text, Text)] -> Ui
choice p os = Choice p os Nothing Nothing

-- | A standalone free-text prompt; 'True' for a multiline (textarea) input.
textIn :: Text -> Bool -> Ui
textIn p m = TextIn p m Nothing Nothing

-- | A FORM-FIELD text input, submitted under @values.<key>@ (used by
-- 'Tidepool.Form'): field key, prompt, multiline flag.
keyedText :: Text -> Text -> Bool -> Ui
keyedText k p m = TextIn p m (Just k) Nothing

-- | A FORM-FIELD text input seeded with an initial (editable) value: field
-- key, prompt, multiline flag, initial text.
keyedTextInitial :: Text -> Text -> Bool -> Text -> Ui
keyedTextInitial k p m initial = TextIn p m (Just k) (Just initial)

-- | A FORM-FIELD radio choice, submitted under @values.<key>@ (used by
-- 'Tidepool.Form'): field key, prompt, (option-key, label) options.
keyedChoice :: Text -> Text -> [(Text, Text)] -> Ui
keyedChoice k p os = Choice p os (Just k) Nothing

-- | A FORM-FIELD radio choice with a pre-selected default option-key.
keyedChoiceSelected :: Text -> Text -> [(Text, Text)] -> Text -> Ui
keyedChoiceSelected k p os sel = Choice p os (Just k) (Just sel)

-- | A FORM-FIELD checkbox group, submitted under @values.<key>@ as an ARRAY
-- of the checked option-keys (used by 'Tidepool.Form'\'s @multiChoiceField@):
-- field key, prompt, (option-key, label) options.
keyedMultiChoice :: Text -> Text -> [(Text, Text)] -> Ui
keyedMultiChoice k p os = MultiChoice p os (Just k)

-- | A small chip (effect-row \/ fan \/ price \/ state).
badge :: Text -> BadgeKind -> Ui
badge = Badge

File: tidepool-harness/src/ui.rs
//! The `Ui` eDSL wire mirror — the contract between the Haskell side
//! (`haskell/lib/Tidepool/Ui.hs`, whose hand-written ToJSON must emit
//! EXACTLY this serde shape) and the tidepool-web renderer. Haskell
//! speaks elicitation semantics; Datastar/HTML vocabulary exists only in
//! the renderer. R0 is first-order (no monadic sequencing — that grows in
//! R1 with the Dialog effect).
//!
//! Every `Choice` elicitation renders with an
//! open-prose escape path IN ADDITION to its options (open sum) — the
//! renderer adds it unconditionally; it is not represented (and not
//! omittable) here.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "ui", rename_all = "snake_case")]
pub enum Ui {
    Card {
        title: String,
        body: Vec<Ui>,
    },
    /// Markdown.
    Prose {
        text: String,
    },
    /// Fenced source block — type signatures, decls, eval sources.
    Code {
        lang: String,
        source: String,
    },
    Choice {
        prompt: String,
        /// (option-key, label) pairs; the answer references the option-key.
        options: Vec<(String, String)>,
        /// FORM FIELD key. When `Some`, this Choice is a field of an enclosing
        /// form: it renders as a radio group whose selection is submitted under
        /// `values.<key>` (not an immediate-post button set). When `None` (the
        /// default, omitted on the wire), it's a standalone one-click choice.
        #[serde(default, skip_serializing_if = "Option::is_none")]
        key: Option<String>,
        /// The option-key pre-selected as the radio group's default.
        #[serde(default, skip_serializing_if = "Option::is_none")]
        selected: Option<String>,
    },
    TextIn {
        prompt: String,
        multiline: bool,
        /// FORM FIELD key. When `Some`, this input is a field of an enclosing
        /// form, submitted under `values.<key>`; when `None` (default, omitted
        /// on the wire), it's a standalone single-value text box.
        #[serde(default, skip_serializing_if = "Option::is_none")]
        key: Option<String>,
        /// The input's starting text — a draft the operator edits.
        #[serde(default, skip_serializing_if = "Option::is_none")]
        initial: Option<String>,
    },
    /// Checkboxes over (option-key, label) pairs — a FORM-FIELD widget, keyed
    /// exactly like a keyed `Choice`: submitted under `values.<key>` as an
    /// ARRAY of the checked option-keys.
    MultiChoice {
        prompt: String,
        options: Vec<(String, String)>,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        key: Option<String>,
    },
    Badge {
        label: String,
        kind: BadgeKind,
    },
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum BadgeKind {
    EffectRow,
    Fan,
    Price,
    State,
}

#[cfg(test)]
mod tests {
    use super::*;

    /// The wire shape is a CROSS-LANGUAGE contract — this test is the
    /// canonical rendering the Haskell ToJSON instances must match
    /// byte-for-byte (module key order, which serde emits in field order).
    #[test]
    fn wire_shape_is_stable() {
        let ui = Ui::Card {
            title: "hole".into(),
            body: vec![
                Ui::Code {
                    lang: "haskell".into(),
                    source: "resume :: Verdict -> M ()".into(),
                },
                Ui::Choice {
                    prompt: "verdict?".into(),
                    options: vec![("approve".into(), "Approve".into())],
                    key: None,
                    selected: None,
                },
                Ui::Badge {
                    label: "Exec, Fs".into(),
                    kind: BadgeKind::EffectRow,
                },
            ],
        };
        let json = serde_json::to_string(&ui).unwrap();
        assert_eq!(
            json,
            r#"{"ui":"card","title":"hole","body":[{"ui":"code","lang":"haskell","source":"resume :: Verdict -> M ()"},{"ui":"choice","prompt":"verdict?","options":[["approve","Approve"]]},{"ui":"badge","label":"Exec, Fs","kind":"effect_row"}]}"#
        );
    }

    /// A form-field `key`, when present, IS emitted (it's how a multi-field
    /// form's submission gets keyed); when absent it's omitted (backward
    /// compat with standalone widgets). Pins both halves of the contract.
    #[test]
    fn keyed_field_wire_shape() {
        let keyed = Ui::TextIn {
            prompt: "Name".into(),
            multiline: false,
            key: Some("f0".into()),
            initial: None,
        };
        assert_eq!(
            serde_json::to_string(&keyed).unwrap(),
            r#"{"ui":"text_in","prompt":"Name","multiline":false,"key":"f0"}"#
        );
        let unkeyed = Ui::TextIn {
            prompt: "Name".into(),
            multiline: false,
            key: None,
            initial: None,
        };
        assert_eq!(
            serde_json::to_string(&unkeyed).unwrap(),
            r#"{"ui":"text_in","prompt":"Name","multiline":false}"#
        );
    }

    /// `TextIn.initial` (a prefill draft) is emitted only when present, same
    /// skip-serialize discipline as `key` — appended AFTER `key` on the wire.
    #[test]
    fn textin_initial_wire_shape() {
        let prefilled = Ui::TextIn {
            prompt: "Notes".into(),
            multiline: false,
            key: Some("f0".into()),
            initial: Some("draft text".into()),
        };
        assert_eq!(
            serde_json::to_string(&prefilled).unwrap(),
            r#"{"ui":"text_in","prompt":"Notes","multiline":false,"key":"f0","initial":"draft text"}"#
        );
        let no_initial = Ui::TextIn {
            prompt: "Notes".into(),
            multiline: false,
            key: Some("f0".into()),
            initial: None,
        };
        assert_eq!(
            serde_json::to_string(&no_initial).unwrap(),
            r#"{"ui":"text_in","prompt":"Notes","multiline":false,"key":"f0"}"#
        );
    }

    /// `Choice.selected` (a pre-selected default) is emitted only when
    /// present, same skip-serialize discipline as `key` — appended AFTER
    /// `key` on the wire.
    #[test]
    fn choice_selected_wire_shape() {
        let with_default = Ui::Choice {
            prompt: "Lane".into(),
            options: vec![("a".into(), "Alpha".into())],
            key: Some("f0".into()),
            selected: Some("a".into()),
        };
        assert_eq!(
            serde_json::to_string(&with_default).unwrap(),
            r#"{"ui":"choice","prompt":"Lane","options":[["a","Alpha"]],"key":"f0","selected":"a"}"#
        );
        let no_default = Ui::Choice {
            prompt: "Lane".into(),
            options: vec![("a".into(), "Alpha".into())],
            key: Some("f0".into()),
            selected: None,
        };
        assert_eq!(
            serde_json::to_string(&no_default).unwrap(),
            r#"{"ui":"choice","prompt":"Lane","options":[["a","Alpha"]],"key":"f0"}"#
        );
    }

    /// `MultiChoice` wire shape, keyed and unkeyed — mirrors `Choice`'s own
    /// keyed/unkeyed discipline.
    #[test]
    fn multichoice_wire_shape() {
        let keyed = Ui::MultiChoice {
            prompt: "Pick some".into(),
            options: vec![("a".into(), "Alpha".into()), ("b".into(), "Beta".into())],
            key: Some("f0".into()),
        };
        assert_eq!(
            serde_json::to_string(&keyed).unwrap(),
            r#"{"ui":"multi_choice","prompt":"Pick some","options":[["a","Alpha"],["b","Beta"]],"key":"f0"}"#
        );
        let unkeyed = Ui::MultiChoice {
            prompt: "Pick some".into(),
            options: vec![("a".into(), "Alpha".into())],
            key: None,
        };
        assert_eq!(
            serde_json::to_string(&unkeyed).unwrap(),
            r#"{"ui":"multi_choice","prompt":"Pick some","options":[["a","Alpha"]]}"#
        );
    }
}
